The Curse Of The Slot Machine - Now Clone-less

san7890 opened the pull request:
## About The Pull Request

Hey there,

I think it's commonly held that clone damage sucks and is overused. One of the last places where it was in slot machine code as a type of "immutable" damage that would cause you to die if you didn't leave and get medical attention. It had a lot of silliness and I'm not sure if a lot of it was meant to work the way it was, but here we are.

So, what's the changes?

* The Cursed Slot Machine will give you a status effect that will "curse" you with repeated damage. After five curses, you get gibbed (similar to the old behavior of the machine). Each curse has it's own change to the status effect, with a lot of depth included. Let me know if the fluff messages about the status effect change sucks, I think it's neat though.
* A person with the curse will smoke. I wanted this to look a bit more "steamy" or grey, but I think it's a decent way of communicating that shit is fucked up with that dude.
* You also get a branded wound after your first failure at the slot machine. Ouchers. Should get it looked at by a doctor.
* We also throw a nice screen alert and all of that jazz.
* I also cleaned up all of the code relating to the slot machine (including a stupid double and), and did some tinkering with the status effects framework to get the desired effect I wanted. I hope you enjoy it as much as I did making it. We use cooldowns and stuff between slot machine pulls.
* If _anyone_ wins on the slot machine, all curses/brands are lifted. Lucky jackpot!
* A lot of the stuff in this code has a lot of vars that might not be modified codewise in case admins still want to jank with this for events.
## Why It's Good For The Game

Clone damage stinks, and I don't really like it as a way to subvert the whole "oh you can't use legion cores to get your way". It's a cursed slot machine, and it should do long term damage so that even if you expend all of the resources on the station, it might all be for naught. It's a horrible price to pay in your search for that d20. I think the negative side effects are pretty OK as far as balance, earlier iterations of this concept had you die _way_ too fast.

All in all, it's just way more of an interesting interaction than "you take damage and then go to medbay and then come back in the hopes of gambling a d20".
## Changelog
:cl:
add: The Lavaland Cursed Slot Machine of Greed suddenly seems a lot more sinister...
refactor: Instead of taking clone damage from the cursed slot machine, you now get a status effect with a number of curses associated with it. There's some interesting florid language associated with the status effect as a nicety until your eventual gibbing from chasing that prize.
/:cl:

I remain undecided if I should keep the curse limit uncapped (but have the damages increase rapidly after the 5-curse threshold), so I left it as the `gib()` from the old code. Let me know your thoughts, but I really don't like the thought that getting the fabled d20 is easy.

san7890 committed:
initial cleanup

san7890 committed:
glingus

san7890 committed:
grlonapwe

san7890 committed:
adsfasdfasd

san7890 committed:
more code fixes

san7890 committed:
nerfs the ability a bit

way too strong

san7890 committed:
i'm going to bed

i'll iron the kinks out in the morning

san7890 committed:
maintains admin schenanigans

san7890 committed:
glarrrg

san7890 committed:
swagger

san7890 committed:
my stommy hort

san7890 committed:
glarg

san7890 committed:
adds special icon for screen alert

san7890 committed:
more fixes

san7890 committed:
upgrades damage prob on stat change

this is more like the old behavior of the old slot machine

san7890 committed:
fixes particle effects

san7890 committed:
has the `use` signal be governed by the slot machine

again i dont' really know how you could get here without admin chicanery but lets be safe

san7890 committed:
that should be on the subtype

san7890 committed:
delete smoke on nongibbed death

CCC23 commented:
damn it now I cant play forever :(

Comxy commented:
> damn it now I cant play forever :(

Don't worry, you can still use slime clones

distributivgesetz commented:
Don't forget 0.1u rezadone pills instahealing your clone so you can continue to gamble without any cost

MrMelbert commented on code/datums/status_effects/debuffs/cursed.dm:
```diff
@@ -0,0 +1,187 @@
+	qdel(src)
+
+/// If our owner's stat changes, rapidly surge the damage chance.
+/datum/status_effect/grouped/cursed/proc/on_stat_changed()
+	if(owner.stat == CONSCIOUS || owner.stat == DEAD) // reset on these two states
```
```suggestion
/datum/status_effect/grouped/cursed/proc/on_stat_changed()
	SIGNAL_HANDLER
	if(owner.stat == CONSCIOUS || owner.stat == DEAD) // reset on these two states
```

MrMelbert commented on code/datums/status_effects/_status_effect.dm:
```diff
@@ -205,3 +205,8 @@
+
+/// Based on what's going on with our status effect and the amount of curses we have, update our description.
+/// Only for subtypes
+/atom/movable/screen/alert/status_effect/proc/update_description()
+	SHOULD_CALL_PARENT(FALSE)
```
Seems a little weird to have this vs just handling it on the status effect itself

san7890 replied:
it seemed like a cleaner implementation to me to have the status effect handle its own desc rather than keep comingling the desc with the status effect and having to keep track of all these different variables

san7890 replied:
using `update_desc()` is a better idea thanks

MrMelbert commented on code/datums/status_effects/debuffs/cursed.dm:
```diff
@@ -0,0 +1,187 @@
+	SIGNAL_HANDLER
+	curse_count++
+
+	if(!isnull(linked_alert))
+		linked_alert.update_description()
```
```suggestion
	linked_alert?.update_description()
```
Work smarter, not harder

san7890 replied:
the question mark is fine for here imo

MrMelbert commented on code/datums/status_effects/debuffs/cursed.dm:
```diff
@@ -0,0 +1,187 @@
+			qdel(src)
+			return
+
+	for(var/message in messages)
+		to_chat(owner, message)
+		sleep(1.5 SECONDS) // yes yes a bit fast but it can be a lot of text and i want the whole thing to send before the cooldown on the slot machine might expire
```
Do we need to worry about them playing it while this is sleeping? 

In a similar vein you may want a sanity check for if the effect is qdel's mid monologue because it'll runtime when owner is set to null

san7890 replied:
no because we cooldown for 15 seconds and this shouldn't exceed 10 seconds. i'll do the qdeleted check just in case though

san7890 replied:
also `to_chat()` nullchecks the target anyways, i don't get what would runtime

![image](https://github.com/tgstation/tgstation/assets/34697715/47ba92a8-e7af-4e0c-b13e-7216a340341b)

Fikou replied:
i would use addtimer to_chat instead

MrMelbert commented on code/datums/status_effects/debuffs/cursed.dm:
```diff
@@ -0,0 +1,187 @@
+
+	if(!isnull(branded_hand))
+		var/datum/wound/brand = branded_hand.get_wound_type(/datum/wound/burn/severe/cursed_brand)
+		if(!isnull(brand))
+			brand.remove_wound()
```
```suggestion
		brand?.remove_wound()
```

MrMelbert commented on code/datums/status_effects/debuffs/cursed.dm:
```diff
@@ -0,0 +1,187 @@
+		return // you get one "freebie" (single damage) to nudge you into thinking this is a bad idea before the house begins to win.
+
+	// the house won.
+	var/ticked_coefficient = rand(0.15, 0.40)
```
Rand does not work with decimals

san7890 replied:
dammit i knew there was something weird with this shit

san7890 committed:
`?` and `SIGNAL_HANDLER`

san7890 committed:
undecimals

san7890 committed:
adds `monologuing` var to ensure that fully sends

san7890 committed:
HOW'D THAT EVEN HAPPEN

san7890 committed:
uses `update_desc()` instead

i put `update_appearance()` there as the invoking proc so people can extend it in the future with minimal concern

san7890 committed:
that comment is useless now

mc-oofert commented:
major greed ruin nerf

san7890 commented:
unfortunately the potential gameplay loop of "get a chemist to make a shitload of drugs so you can bypass the soft gambling limit" in the hopes of getting a further chance of becoming a wizard or something like that through a d20 seems a bit overpowered to me and i'm not fond of it. also it was boring

san7890 committed:
Merge remote-tracking branch 'upstream/master' into fasdfasdf

san7890 committed:
updates wound descriptions

AtomTheProphet commented:
> unfortunately the potential gameplay loop of "get a chemist to make a shitload of drugs so you can bypass the soft gambling limit" in the hopes of getting a further chance of becoming a wizard or something like that through a d20 seems a bit overpowered to me and i'm not fond of it. also it was boring

This change gets rid of that loop and adds interesting long term consequences. Very good!

Arturlang commented:
Is 5 times the same amount you could do with the old machine?

san7890 commented:
> Is 5 times the same amount you could do with the old machine without healing?

In theory, yes, because it applied 20 Clone Damage every time you used it (regardless of success or failure). Since you hit crit at 100 of any damage, you would have had only five uses of it at the time this was implemented (before chemistry and other things got overhauled to the point where clone damage became more and more trivial to heal in response to other things dealing clone damage).

JohnFulpWillard commented on code/modules/mapfluff/ruins/objects_and_mobs/cursed_slot_machine.dm:
```diff
@@ -0,0 +1,121 @@
+
+	var/mob/living/carbon/human/human_user = user
+
+	if(obj_flags & IN_USE || !COOLDOWN_FINISHED(src, spin_cooldown))
```
can you remove the use of ``IN_USE`` here :)

san7890 replied:
Okay

Rhials commented on code/modules/mapfluff/ruins/objects_and_mobs/cursed_slot_machine.dm:
```diff
@@ -0,0 +1,121 @@
+	/// The maximum amount of curses we will allow a player to have before disallowing them to use the machine.
+	var/max_curse_amount = 5
+	/// clone damaged dealt each roll
+	var/damage_on_roll = 20
```
Unused var? This doesn't do clone damage anymore right?

san7890 replied:
Good point

san7890 committed:
deletes unused var

thanks rhials duffy

san7890 committed:
moves in_use to a var

ATH1909 commented on code/datums/status_effects/debuffs/cursed.dm:
```diff
@@ -0,0 +1,193 @@
+
+		if(3)
+			owner.emote("cough")
+			messages += span_boldwarning("Your throat becomes to feel like it's slowly caking up with sand and dust. You eject the contents of the back of your throat onto your good hand.")
```
what if you only have one arm?

san7890 committed:
wamp wamp

Fikou commented on code/datums/status_effects/debuffs/cursed.dm:
```diff
@@ -0,0 +1,193 @@
+	switch(curse_count)
+		if(2 to 3)
+			particle_path = /particles/smoke/steam
+		if(4)
```
```suggestion
		if(4 to INFINITY)
```

san7890 committed:
oh right i forgot that should be infinite too

Co-authored-by: Fikou <23585223+Fikou@users.noreply.github.com>
